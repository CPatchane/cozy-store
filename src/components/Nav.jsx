import React from 'react'
import { translate } from 'cozy-ui/react/I18n'
import { NavLink } from 'react-router-dom'

export const Nav = ({ t }) => (
  <nav>
    <ul className='coz-nav'>
      <li className='coz-nav-item'>
        <NavLink to='/discover' className='sto-cat-discover coz-nav-link' activeClassName='active'>
          { t('nav.discover') }
        </NavLink>
      </li>
      <li className='coz-nav-item'>
        <NavLink to='/myapps' className='sto-cat-myapps coz-nav-link' activeClassName='active'>
          { t('nav.myapps') }
        </NavLink>
      </li>
    </ul>
  </nav>
)

export default translate()(Nav)
