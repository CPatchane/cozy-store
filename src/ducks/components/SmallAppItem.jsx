import React from 'react'

import Icon from 'cozy-ui/react/Icon'
import { translate } from 'cozy-ui/react/I18n'

import defaultAppIcon from '../../assets/icons/icon-cube.svg'
import { Link } from 'react-router-dom'
import Button from 'cozy-ui/react/Button'

export const SmallAppItem = ({
  t,
  slug,
  developer,
  editor,
  icon,
  name,
  namePrefix,
  installed,
  onClick,
  installedAppLink
}) => {
  const onShortcutAppButton = (e, link) => {
    // prevent from opening parent SmallAppItem link
    // which will lead to the application main page
    e.stopPropagation()
    if (link) window.location.assign(link)
  }
  return (
    // HACK a11y
    // `onKeyDown={(e) => e.keyCode === 13 ? onClick() : null`
    // This syntax make the div accessible.
    <div
      className='sto-small-app-item'
      onKeyDown={e => (e.keyCode === 13 ? onClick() : null)}
      onClick={onClick}
      tabIndex={0}
    >
      {icon ? (
        <img
          src={icon}
          alt={`${slug}-icon`}
          width='64'
          height='64'
          className='sto-small-app-item-icon'
        />
      ) : (
        <Icon
          className='sto-small-app-item-icon--default blurry'
          width='48'
          height='64'
          icon={defaultAppIcon}
          color='#95999D'
        />
      )}
      {installed && (
        <Icon
          className='sto-small-app-item-badge-installed'
          icon='check'
          color='#2bba40'
        />
      )}
      <div className='sto-small-app-item-desc'>
        <h4 className='sto-small-app-item-title'>
          {namePrefix && `${namePrefix} `}{name}
        </h4>
        <p className='sto-small-app-item-detail'>
          {developer.name === 'Cozy' ? 'Cozy Cloud Inc.' : developer.name}
        </p>
        {installed
          ? <Button
            onClick={(e) => onShortcutAppButton(e, installedAppLink)}
            label={t('app_item.open')}
            theme='secondary'
            className='sto-small-app-item-button'
            size='tiny'
          />
          : <Link
            to={`/discover/${slug}/manage`}
            onClick={(e) => onShortcutAppButton(e)}
            className='c-btn c-btn--regular c-btn--tiny sto-small-app-item-button sto-small-app-item-button-install'
          >
            <span>{t('app_item.install')}</span>
          </Link>
        }
      </div>
    </div>
  )
}

export default translate()(SmallAppItem)
