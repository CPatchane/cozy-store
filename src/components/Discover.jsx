import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'

import SmallAppItem from './SmallAppItem'

const DEMO_APPS = [
  {
    slug: 'bank',
    name: 'Bank',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-bank.svg')
  },
  {
    slug: 'sante',
    name: 'Sant√©',
    developer: {name: 'MAIF Ltd'},
    editor: '',
    icon: require('../assets/temp_apps/icon-sante.svg')
  },
  {
    slug: 'mon_logis',
    name: 'Mon Logis',
    developer: {name: 'HoodBrains'},
    editor: '',
    icon: require('../assets/temp_apps/icon-mon_logis.svg')
  },
  {
    slug: 'calendar',
    name: 'Calendar',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-calendar.svg')
  }
]

class Discover extends Component {
  render () {
    const { t } = this.props
    return (
      <div className='sto-discover sto-soon'>
        <h2 className='sto-discover-title'>{t('discover.title')}</h2>
        <div className='sto-discover-get-started'>
          <h3 className='sto-discover-get-started-title'>
            {t('discover.get_started')}
          </h3>
          <div className='sto-discover-get-started-list'>
            {DEMO_APPS.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
              />
            })}
          </div>
        </div>
      </div>
    )
  }
}

export default translate()(Discover)
