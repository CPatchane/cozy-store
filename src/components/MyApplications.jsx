import React, { Component } from 'react'

import { Route } from 'react-router-dom'
import { translate } from 'cozy-ui/react/I18n'
import Spinner from 'cozy-ui/react/Spinner'

import SmallAppItem from './SmallAppItem'
import ApplicationModal from './ApplicationModal'

class MyApplications extends Component {
  constructor (props) {
    super(props)
    props.fetchApps()
  }

  onAppClick (appSlug) {
    this.props.history.push(`/myapps/${appSlug}/manage`)
  }

  render () {
    const { t, myApps, isFetching, error } = this.props
    return (
      <div className='sto-myapps'>
        <h2 className='sto-myapps-title'>{t('myapps.title')}</h2>
        <div className='sto-myapps-list'>
          {!isFetching && myApps && !!myApps.length &&
            myApps.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
                onClick={() => this.onAppClick(a.slug)}
              />
            })
          }
          {error &&
            <p>{error}</p>
          }
        </div>
        {isFetching &&
          <Spinner
            size='xxlarge'
            loadingType='appsFetching'
            middle='true'
          />
        }

        <Route path='/myapps/:appSlug/manage' render={({ match }) =>
          (<ApplicationModal {...this.props} match={match} />)
        } />
      </div>
    )
  }
}

export default translate()(MyApplications)
