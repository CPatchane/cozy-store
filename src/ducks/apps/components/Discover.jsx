import React, { Component } from 'react'

import { Route } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import Spinner from 'cozy-ui/react/Spinner'

import SmallAppItem from '../../components/SmallAppItem'
import UninstallModal from './UninstallModal'
import { InstallModal } from '../currentAppVersion/Containers'

export class Discover extends Component {
  constructor (props, context) {
    super(props)
    props.fetchApps(context.lang)

    this.onAppClick = this.onAppClick.bind(this)
  }

  onAppClick (appSlug) {
    this.props.history.push(`/discover/${appSlug}/manage`)
  }

  render () {
    const { t, apps, isFetching } = this.props
    return (
      <div className='sto-discover'>
        <h2 className='sto-discover-title'>{t('discover.title')}</h2>
        <div className='sto-discover-get-started'>
          <h3 className='sto-discover-get-started-title'>
            {t('discover.get_started')}
          </h3>
          <div className='sto-discover-get-started-list'>
            {!isFetching && apps.map(a => {
              // don't display app installed but not in registry
              if (!a.isInRegistry) return null
              let version = a.version
              if (a.versions && Array.isArray(a.versions.stable)) {
                version = (a.versions.stable && a.versions.stable[a.versions.stable.length - 1])
              }
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer || ''}
                editor={a.editor || ''}
                icon={a.icon}
                name={a.name}
                version={version}
                installed={a.installed}
                onClick={() => this.onAppClick(a.slug)}
              />
            })}
            {isFetching &&
              <Spinner
                size='xxlarge'
                loadingType='appsFetching'
                middle='true'
              />
            }
          </div>
        </div>
        <Route path='/discover/:appSlug/manage' render={({ match }) => {
          if (isFetching) return
          if (apps.length && match.params) {
            const app = apps.find(app => app.slug === match.params.appSlug)
            if (app.installed) {
              return <UninstallModal uninstallApp={this.props.uninstallApp} parent='/discover' app={app} match={match} />
            } else if (!app.installed) {
              return <InstallModal installApp={this.props.installApp} currentAppVersion={this.props.currentAppVersion} parent='/discover' app={app} match={match} />
            }
          }
        }} />
      </div>
    )
  }
}

export default translate()(Discover)
