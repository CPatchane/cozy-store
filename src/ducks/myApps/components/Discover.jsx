import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'
import Spinner from 'cozy-ui/react/Spinner'

import SmallAppItem from '../../components/SmallAppItem'

export class Discover extends Component {
  constructor (props) {
    super(props)
    props.fetchApps()
  }

  render () {
    const { t, apps, isFetching } = this.props
    return (
      <div className='sto-discover'>
        <h2 className='sto-discover-title'>{t('discover.title')}</h2>
        <div className='sto-discover-get-started'>
          <h3 className='sto-discover-get-started-title'>
            {t('discover.get_started')}
          </h3>
          <div className='sto-discover-get-started-list'>
            {!isFetching && apps.map(a => {
              const version = a.version || (a.versions && a.versions.stable && a.versions.stable[a.versions.stable.length - 1]) || ''
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer || ''}
                editor={a.editor || ''}
                icon={a.icon}
                name={a.name}
                version={version}
                installed={a.installed}
              />
            })}
            {isFetching &&
              <Spinner
                size='xxlarge'
                loadingType='appsFetching'
                middle='true'
              />
            }
          </div>
        </div>
      </div>
    )
  }
}

export default translate()(Discover)
