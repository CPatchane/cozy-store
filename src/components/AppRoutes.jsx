import React from 'react'
import { Route, Switch, Redirect } from 'react-router-dom'

import MyApplications from '../containers/MyApplications'

const AppRoutes = () => {
  return <Switch>
    <Route path='/myapps' component={MyApplications} />
    <Redirect from='/' to='/myapps' />
  </Switch>
}

export default AppRoutes
