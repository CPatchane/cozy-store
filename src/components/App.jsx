
import React from 'react'

import Sidebar from './Sidebar'
import AppRoutes from './AppRoutes'

const App = () => (
  <div className='sto-wrapper coz-sticky'>
    <Sidebar />
    <main className='sto-content'>
      <AppRoutes />
    </main>
  </div>
)
export default App
