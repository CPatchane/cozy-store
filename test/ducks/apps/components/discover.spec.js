'use strict'

/* eslint-env jest */

import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { tMock } from '../../../jestLib/I18n'
import SmallAppItem from 'ducks/components/SmallAppItem'
import { Discover } from 'ducks/apps/components/Discover'

import mockApps from '../_mockApps'

Enzyme.configure({ adapter: new Adapter() })

const mockError = new Error('This is a test error')

const mockRegistyApps = mockApps.filter(app => app.isInRegistry).filter(app =>
(Array.isArray(app.versions.stable) && !!app.versions.stable.length))

const getMockProps = (apps = mockRegistyApps, isFetching = false, fetchError = null, match = { isExact: true }) => ({
  fetchApps: jest.fn(),
  apps,
  isFetching,
  fetchError,
  history: { push: jest.fn() },
  match
})

describe('Discover component', () => {
  it('should be rendered correctly with apps', () => {
    const mockProps = getMockProps()
    const component = shallow(
      <Discover t={tMock} {...mockProps} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render correctly a spinner if apps is fetching', () => {
    const mockProps = getMockProps([], true, null)
    const component = shallow(
      <Discover t={tMock} {...mockProps} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should display error from props correctly', () => {
    const mockProps = getMockProps([], false, mockError)
    const component = shallow(
      <Discover t={tMock} {...mockProps} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should not render apps list if !match.isExact', () => {
    const mockProps = getMockProps(mockRegistyApps, false, null, { isExact: false })
    const component = shallow(
      <Discover t={tMock} {...mockProps} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should handle correctly items onClick', () => {
    const mockProps = getMockProps()
    const component = shallow(
      <Discover t={tMock} {...mockProps} />
    )
    expect(component.find(SmallAppItem).length).toBe(mockRegistyApps.length)
    const appItem = component.find(SmallAppItem).at(0).dive() // shallow on more level on first app item
    appItem.simulate('click')
    // history push to app modal URL
    expect(mockProps.history.push.mock.calls.length).toBe(1)
    expect(mockProps.history.push.mock.calls[0][0]).toBe(`/discover/${mockRegistyApps[0].slug}`)
  })
})
