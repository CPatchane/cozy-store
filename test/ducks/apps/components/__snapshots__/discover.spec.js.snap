// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discover component should be rendered correctly with apps 1`] = `
<div
  className="sto-discover"
>
  <div>
    <h2
      className="sto-discover-title"
    >
      Discover
    </h2>
    <div
      className="sto-discover-sections"
    >
      <Wrapper
        apps={
          Array [
            Object {
              "developer": Object {
                "name": "Cozy",
              },
              "editor": "Cozy",
              "icon": "<svg></svg>",
              "installed": true,
              "isInRegistry": true,
              "name": "Collect",
              "name_prefix": "Cozy",
              "related": "http://collect.cozy.mock/",
              "slug": "collect",
              "type": "webapp",
              "uninstallable": false,
              "version": "3.0.0",
              "versions": Object {
                "beta": Array [
                  "3.0.0",
                ],
                "dev": Array [
                  "3.0.0",
                ],
                "stable": Array [
                  "3.0.0",
                ],
              },
            },
            Object {
              "developer": Object {
                "name": "Cozy",
              },
              "icon": "<svg></svg>",
              "isInRegistry": true,
              "name": "Trinlane",
              "slug": "konnector-trinlane",
              "type": "konnector",
              "uninstallable": true,
              "versions": Object {
                "beta": Array [
                  "0.1.0",
                ],
                "dev": Array [
                  "0.1.0",
                ],
                "stable": Array [
                  "0.1.0",
                ],
              },
            },
            Object {
              "developer": Object {
                "name": "Cozy",
              },
              "editor": "Cozy",
              "icon": "<svg></svg>",
              "isInRegistry": true,
              "name": "Photos",
              "name_prefix": "Cozy",
              "slug": "photos",
              "type": "webapp",
              "uninstallable": true,
              "versions": Object {
                "beta": Array [
                  "3.0.0",
                ],
                "dev": Array [
                  "3.0.0",
                ],
                "stable": Array [
                  "3.0.0",
                ],
              },
            },
            Object {
              "icon": "<svg></svg>",
              "isInRegistry": true,
              "name": "Tasky",
              "slug": "tasky",
              "type": "webapp",
              "uninstallable": true,
              "versions": Object {
                "beta": Array [
                  "1.0.0",
                ],
                "dev": Array [
                  "1.0.0",
                ],
                "stable": Array [
                  "1.0.0",
                ],
              },
            },
          ]
        }
        error={null}
        onAppClick={[Function]}
      />
    </div>
  </div>
  <withRouter(ApplicationRouting)
    actionError={null}
    apps={
      Array [
        Object {
          "developer": Object {
            "name": "Cozy",
          },
          "editor": "Cozy",
          "icon": "<svg></svg>",
          "installed": true,
          "isInRegistry": true,
          "name": "Collect",
          "name_prefix": "Cozy",
          "related": "http://collect.cozy.mock/",
          "slug": "collect",
          "type": "webapp",
          "uninstallable": false,
          "version": "3.0.0",
          "versions": Object {
            "beta": Array [
              "3.0.0",
            ],
            "dev": Array [
              "3.0.0",
            ],
            "stable": Array [
              "3.0.0",
            ],
          },
        },
        Object {
          "developer": Object {
            "name": "Cozy",
          },
          "icon": "<svg></svg>",
          "isInRegistry": true,
          "name": "Trinlane",
          "slug": "konnector-trinlane",
          "type": "konnector",
          "uninstallable": true,
          "versions": Object {
            "beta": Array [
              "0.1.0",
            ],
            "dev": Array [
              "0.1.0",
            ],
            "stable": Array [
              "0.1.0",
            ],
          },
        },
        Object {
          "developer": Object {
            "name": "Cozy",
          },
          "editor": "Cozy",
          "icon": "<svg></svg>",
          "isInRegistry": true,
          "name": "Photos",
          "name_prefix": "Cozy",
          "slug": "photos",
          "type": "webapp",
          "uninstallable": true,
          "versions": Object {
            "beta": Array [
              "3.0.0",
            ],
            "dev": Array [
              "3.0.0",
            ],
            "stable": Array [
              "3.0.0",
            ],
          },
        },
        Object {
          "icon": "<svg></svg>",
          "isInRegistry": true,
          "name": "Tasky",
          "slug": "tasky",
          "type": "webapp",
          "uninstallable": true,
          "versions": Object {
            "beta": Array [
              "1.0.0",
            ],
            "dev": Array [
              "1.0.0",
            ],
            "stable": Array [
              "1.0.0",
            ],
          },
        },
      ]
    }
    installApp={undefined}
    isFetching={false}
    isInstalling={undefined}
    parent="discover"
    uninstallApp={undefined}
  />
</div>
`;

exports[`Discover component should display error from props correctly 1`] = `
<div
  className="sto-discover"
>
  <div>
    <h2
      className="sto-discover-title"
    >
      Discover
    </h2>
    <div
      className="sto-discover-sections"
    >
      <Wrapper
        apps={Array []}
        error={[Error: This is a test error]}
        onAppClick={[Function]}
      />
    </div>
  </div>
  <withRouter(ApplicationRouting)
    actionError={null}
    apps={Array []}
    installApp={undefined}
    isFetching={false}
    isInstalling={undefined}
    parent="discover"
    uninstallApp={undefined}
  />
</div>
`;

exports[`Discover component should not render apps list if !match.isExact 1`] = `
<div
  className="sto-discover"
>
  <withRouter(ApplicationRouting)
    actionError={null}
    apps={
      Array [
        Object {
          "developer": Object {
            "name": "Cozy",
          },
          "editor": "Cozy",
          "icon": "<svg></svg>",
          "installed": true,
          "isInRegistry": true,
          "name": "Collect",
          "name_prefix": "Cozy",
          "related": "http://collect.cozy.mock/",
          "slug": "collect",
          "type": "webapp",
          "uninstallable": false,
          "version": "3.0.0",
          "versions": Object {
            "beta": Array [
              "3.0.0",
            ],
            "dev": Array [
              "3.0.0",
            ],
            "stable": Array [
              "3.0.0",
            ],
          },
        },
        Object {
          "developer": Object {
            "name": "Cozy",
          },
          "icon": "<svg></svg>",
          "isInRegistry": true,
          "name": "Trinlane",
          "slug": "konnector-trinlane",
          "type": "konnector",
          "uninstallable": true,
          "versions": Object {
            "beta": Array [
              "0.1.0",
            ],
            "dev": Array [
              "0.1.0",
            ],
            "stable": Array [
              "0.1.0",
            ],
          },
        },
        Object {
          "developer": Object {
            "name": "Cozy",
          },
          "editor": "Cozy",
          "icon": "<svg></svg>",
          "isInRegistry": true,
          "name": "Photos",
          "name_prefix": "Cozy",
          "slug": "photos",
          "type": "webapp",
          "uninstallable": true,
          "versions": Object {
            "beta": Array [
              "3.0.0",
            ],
            "dev": Array [
              "3.0.0",
            ],
            "stable": Array [
              "3.0.0",
            ],
          },
        },
        Object {
          "icon": "<svg></svg>",
          "isInRegistry": true,
          "name": "Tasky",
          "slug": "tasky",
          "type": "webapp",
          "uninstallable": true,
          "versions": Object {
            "beta": Array [
              "1.0.0",
            ],
            "dev": Array [
              "1.0.0",
            ],
            "stable": Array [
              "1.0.0",
            ],
          },
        },
      ]
    }
    installApp={undefined}
    isFetching={false}
    isInstalling={undefined}
    parent="discover"
    uninstallApp={undefined}
  />
</div>
`;

exports[`Discover component should render correctly a spinner if apps is fetching 1`] = `
<div
  className="sto-discover"
>
  <div>
    <h2
      className="sto-discover-title"
    >
      Discover
    </h2>
    <div
      className="sto-discover-sections"
    />
  </div>
  <withRouter(ApplicationRouting)
    actionError={null}
    apps={Array []}
    installApp={undefined}
    isFetching={true}
    isInstalling={undefined}
    parent="discover"
    uninstallApp={undefined}
  />
  <Wrapper
    loadingType="appsFetching"
    middle="true"
    size="xxlarge"
  />
</div>
`;
